src/h3lib/lib/h3Index.c

```diff
@@ -734,3 +760,232 @@ void H3_EXPORT(h3ToGeoBoundary)(H3Index h3, GeoBoundary* gb) {
  *         a Class II grid.
  */
 int isResClassIII(int res) { return res % 2; }
+
+/**
+ * Produces ijk+ coordinates for an index anchored by an origin.
+ *
+ * The coordinate space used by this function may have deleted
+ * regions or warping due to pentagonal distortion.
+ *
+ * Coordinates are only comparable if they come from the same
+ * origin index.
+ *
+ * @param origin An anchoring index for the ijk+ coordinate system.
+ * @param index Index to find the coordinates of
+ * @param out ijk+ coordinates of the index will be placed here on success
+ * @return 0 on success, or another value on failure.
+ */
+int h3ToIjk(H3Index origin, H3Index h3, CoordIJK* out) {
+    if (H3_GET_MODE(origin) != H3_HEXAGON_MODE ||
+        H3_GET_MODE(h3) != H3_HEXAGON_MODE) {
```

Comment:
Nit: I'm not sure if this is worthwhile. We technically ought to be doing this check on every H3 index algo, and then there's a slippery slope about what we ought to be checking - e.g. would `h3IsValid` be better here? I have mixed feelings about this, since it's basically a service to the developer to guard against stuff we can't type-check, but because it has perf issues in some cases, we can only do it in some functions and not others. It would be nice if we had a consistent approach to this across all algo functions.

The ideal option would be to make this a compile-time check. I don't see any way to do that with externally-provided indexes though.

Reply:
Using assertions makes sense if you have people using debug builds versus release builds. Other languages tend to default to release builds so that is still an issue.

Reply:
I think some of the other algos don't do this because they do not have a way to return error cases to the caller. Other functions like geoToH3 do check for out of bounds inputs and return 0, and geoToH3 is certainly performance critical.

I would rather not use assertions, especially in release builds because instead of propagating an error it would abort the entire process, which would probably be bad.

Reply:
Assertions are not compiled into release builds. The definition of `assert` looks something like this:

```
#ifdef NDEBUG  /* Release mode */
#define assert(cond) /* Empty definition */
#else  /* Debug mode */
#define assert(cond) ...  /* < Define assert macro */
#endif
```